Node Manager: A Few Minor UX Enhancements

jacderida opened the pull request:
- e247ef79 **fix: apply interval only to non-running nodes**

  The node manager `start` command has an `--interval` argument that can be used as a time-based delay
  between starting nodes. The problem is, the interval would apply to nodes that were already running,
  which is a waste of time. Now the interval is not applied to those nodes.

- be04805a **chore: use better banners**

  These new banners for the node manager are a bit more eye catching, and also fix the incorrect sizes
  of the previous banners in the `status --details` command.

- 8a0b296d **chore: output reward balance in `status --json` cmd**

  This was a community feedback request.

  The reward balance is now output as part of the `status --json` command, which involved adding the
  field to the information tracked by the registry.

  BREAKING CHANGE: previously written registry files will not have this new field and so won't
  deserialize correctly.
  
- 5a323651 **tests: use node registry for status**

  Rather than parsing the text-based output of the `status` command, we now switch to using `status
  --json`, which we can then serialize into the `NodeRegistry`, and any status can be obtained from
  there.

  This prevents breakage when the text output changes.
  
- 5e801a0f **fix: change reward balance to optional**

  Due to it being possible to run the node manager `status` command before any services have been
  started, the reward balance should be an optional value, since the node's wallet has not been
  created before it starts.
  
- 94ef20e4 **test: disable node man integration tests**

  These tests are no longer passing and it's not completely clear why. In any case, I think they need
  to be a bit more isolated, and operate on their own local network.

  At the moment there is not enough time to address that.  

## Description

reviewpad:summary

joshuef commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -443,19 +469,42 @@ impl<'a> Joiner<'a> {
+            match msg {
+                NodeMsg::JoinResponse(resp) => return Ok((*resp, sender)),
+                _ => {
+                    trace!("Bootstrap message discarded: sender: {sender:?} msg: {msg:?}")
```
I wonder if this log is clearer as 
```suggestion
                    trace!("Non-Bootstrap message received and discarded: sender: {sender:?} msg: {msg:?}")
```

joshuef commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -443,19 +469,42 @@ impl<'a> Joiner<'a> {
-                    trace!("Bootstrap message discarded: sender: {sender:?} msg_id: {msg_id:?}");
+                } => return Ok(section_tree_update),
+                _ => {
+                    trace!("Bootstrap message discarded: sender: {sender:?} msg: {msg:?}")
```
same thing here. (and perhaps something to distinguish between the two discard cases?)

joshuef reviewed (commented):
non critical, just something that's bugged me whenever I've poked about this msg previously

maqi commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -443,19 +469,42 @@ impl<'a> Joiner<'a> {
+    #[tracing::instrument(skip(self))]
+    async fn receive_node_msg(&mut self) -> Result<(NodeMsg, Peer)> {
+        while let Some(MsgFromPeer { sender, wire_msg }) = self.incoming_msgs.recv().await {
+            // We are interested only in `JoinResponse` type of messages
```
only interested in `Node` type of messages?

maqi commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -357,6 +360,29 @@ impl<'a> Joiner<'a> {
+                    .await
+                    .map_err(|_| Error::JoinTimeout)??;
+
+            if !self.section_tree.update(section_tree_update)? {
```
better put a comment here that: keeps fetching section_tree as long as there is an update being carried out

Meanwhile, this is one section_tree from one node, if the update returns with `updated` , if means we will send a probe to `All` elders again? which is abuse of messaging?

Also, if the update returns with an error, the loop will be terminated too early? i.e. we shall wait for other elders' section_tree as well?

davidrusu replied:
I've reworked the logic here a little bit. We now wait until we receive a threshold of updates before we check for new information.

The only error that would terminate the loop early is if our connections are closed or if we timeout waiting for a section tree update.

davidrusu commented on sn_interface/src/network_knowledge/sections_dag.rs:
```diff
@@ -205,6 +205,17 @@ impl SectionsDAG {
     pub fn partial_dag(&self, from: &bls::PublicKey, to: &bls::PublicKey) -> Result<Self> {
         // start from the "to" key (bottom of the tree) and traverse to the root
         let mut crdt_ops: Vec<Node<SectionInfo>> = Vec::new();
+
+        // Return the singleton DAG when from == to == genesis
+        // This is a special case because the genesis entry does not have a `Node` so `get_node(to)` will fail
+        if to == &self.genesis_key {
+            if from != &self.genesis_key {
+                return Err(Error::InvalidBranch);
+            } else {
+                return Ok(Self::new(self.genesis_key));
+            }
+        }
+
```
cc @RolandSherwin small change to `SectionsDag::partial_dag` semantics here.

maqi commented on sn_node/src/node/messaging/node_msgs.rs:
```diff
@@ -196,10 +193,11 @@ impl MyNode {
+                            let section_tree_update = node
+                                .network_knowledge
+                                .section_tree()
+                                .generate_section_tree_update(&Prefix::default())?; // TODO: remove this dummy update
```
I'd prefer this TODO to be resolved within this PR.
and using a `dummy` update could incur unexpected issue.
Maybe a quick one is just to just change the constructor to:
```rust
NetworkKnowledge::new(SectionTree, Option<SectionTreeUpdate>)
```

davidrusu replied:
I tried quite a bit today to make this work. The problem is that we need a signed_sap for the network_knowledge and it's currently being retrieved from `section_tree_update.signed_sap`.

Any refactors I did spiralled into much bigger changes.

How strong is your preference to get this resolved before merging?

maqi replied:
Ok for leave it for a follow-up PR, as long as the comment can make it clear using a `dummy` here is a really `temp` and shall have no critical impact.
and thx for clarify it.

davidrusu replied:
I'll make this a ticket so we don't forget about it.

davidrusu replied:
#1798

maqi commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -274,40 +229,22 @@ impl<'a> Joiner<'a> {
                         continue;
```
This may not related to this PR, however, as I understand,
the above `if !section_auth.prefix().matches(&self.node.name())` shall use the `section_auth` of the received msg, 
not the `section_auth of JoinResponse::Redirect(section_auth)` .
As the node could be redirected to a different section?

davidrusu replied:
Then what would be the purpose of the section-auth in the `JoinResponse::Redirect(section_auth)`?

My understanding here is that we are not receiving this response from the redirect target section, we are receiving it from the elders we had just sent the request to.

maqi replied:
>Then what would be the purpose of the section-auth in the JoinResponse::Redirect(section_auth)

At https://github.com/maidsafe/safe_network/blob/main/sn_node/src/node/bootstrap/join.rs#L300 we use the elders of the received section_auth within the Redicrect as the new elders target, and at https://github.com/maidsafe/safe_network/blob/main/sn_node/src/node/messaging/join.rs#L57 it's obvious that the `sap of the closest to the target section` we wrap within the Redirect message.

This check might get wrong due to a copy/paste from Retry branch ?
another possibility could be this check is to resolve the issue when the node got redirected to the self section.
However, a resolvement like this will be problematic if the node get redirected to other section.

davidrusu replied:
My reading of that check was to ensure that this joining node's name fits within that target section we are being redirected to. (Recall that the join logic doesn't regenerate a name on redirect, names are only regenerated on retries)

Another option would be to always regenerate names, then I think we don't need to keep this prefix check since we'll regenerate the name for the target section's prefix.

maqi commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -274,40 +229,22 @@ impl<'a> Joiner<'a> {
 
-                    section_key = new_section_key;
-                    self.prefix = section_auth.prefix();
+                    self.bootstrap_section_tree(section_auth, response_timeout)
```
just to verify:
this means for each individual redirect/retry response, we will startup a `bootstrap_section_tree` round, which means sending messages to all elders?
That will cause un-necessary messages and may even a message storming?

davidrusu replied:
Not quite, the retry cache prevents us from re-bootstrapping until we have t + 1 retries.

The redirect logic did not have that cache in front of it so it was doing a join request for every redirect response. The flow remains the same with this refactor, the difference is that we add a bootstrap phase in-front of the join request.

I'd prefer to keep the logic as it was before i.e. no cache for redirect as that's an orthogonal change to this PR. We can make the change in a followup PR.

maqi commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -274,40 +229,22 @@ impl<'a> Joiner<'a> {
+                        .await?;
 
-                    let msg = JoinRequest { section_key };
+                    let target_sap = self.join_target_sap()?;
```
when a new node get redirected, it actually need to re-generate it's name to match the target section?

davidrusu replied:
Perhaps it should, the retry logic re-generates the name, so maybe redirect should as well.

But here as well, that change seems orthogonal, we can fix this in another PR since the existing behaviour does not re-generate names.

maqi commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -321,6 +258,48 @@ impl<'a> Joiner<'a> {
+
+            for _ in 0..target_sap.public_key_set().threshold() {
+                let update =
+                    tokio::time::timeout(response_timeout, self.receive_section_tree_update())
```
the passed in `response_timeout` is actually the overall `join_time_out` 
but here the work to be carried out is just for `query  one elder` , and we only start a new one AFTER the previous one completed or timedout. 
which means whenever we encountered the first timeout here, the overall timeout is reached.
In normal case, this only leaves us ONE chance to just query the first elder?

maybe we shall use `join_time_out / elders.len()` as per action timeout, or make the actions to be undertaken parrallellized?

further more, the `receive_section_tree_update` function itself is a `looping receiver` , 
whichmeans the `for` clause here is bit confusing/wasted or even incorrect?

davidrusu replied:
Not quite, the timeout passed in is `join_time_out / 10` (see `try_join(..)` definition.

And the timeout here is to receive *any* section tree update. We've sent an AEProbe broadcast above to all elders, we are now waiting for any of them to respond.

The responses are all coming in in parallel, they get queued in the `mpsc::channel`, and `self.receive_section_tree_update()` simply waits on that channel until it sees a section tree update. They will be processed as quickly as they are received. (these are not serial requests to each elder).

> further more, the receive_section_tree_update function itself is a looping receiver , whichmeans the for clause here is bit confusing/wasted or even incorrect?

The for loop here is saying, "wait for a threshold of section tree updates to come in". The `received_section_tree_update()` waits for a single section tree update to come in.

Let me know if anything here is unclear.

maqi commented on sn_node/src/node/bootstrap/join.rs:
```diff
@@ -321,6 +258,48 @@ impl<'a> Joiner<'a> {
+                target_sap = self.join_target_sap()?;
+            } else {
+                // We are up to date with these nodes so we can end the bootstrap
+                return Ok(());
```
the default value of `any_new_information ` is `false`, which means if all timedout, we will still return as `OK`.
Currently, the overall timeout is the only way to bail us out. However this could be tricky/dangerous.

davidrusu replied:
The any timeout in the for-loop will bail out (note the ??) in:

```rust
                    tokio::time::timeout(response_timeout, self.receive_section_tree_update())
                        .await
                        .map_err(|_| Error::JoinTimeout)??;
```

davidrusu commented:
bors r+

jacderida committed:
fix: apply interval only to non-running nodes

The node manager `start` command has an `--interval` argument that can be used as a time-based delay
between starting nodes. The problem is, the interval would apply to nodes that were already running,
which is a waste of time. Now the interval is not applied to those nodes.

jacderida committed:
chore: use better banners

These new banners for the node manager are a bit more eye catching, and also fix the incorrect sizes
of the previous banners in the `status --details` command.

jacderida committed:
chore: output reward balance in `status --json` cmd

This was a community feedback request.

The reward balance is now output as part of the `status --json` command, which involved adding the
field to the information tracked by the registry.

BREAKING CHANGE: previously written registry files will not have this new field and so won't
deserialize correctly.

jacderida committed:
test: use node registry for status

Rather than parsing the text-based output of the `status` command, we now switch to using `status
--json`, which we can then serialize into `StatusSummary`, and any status can be obtained from
there.

This prevents breakage when the text output changes.

jacderida committed:
fix: change reward balance to optional

Due to it being possible to run the node manager status command before any services have been
started, the reward balance should be an optional value, since the node's wallet has not been
created before it starts.

jacderida committed:
test: disable node man integration tests

These tests are no longer passing and it's not completely clear why. In any case, I think they need
to be a bit more isolated, and operate on their own local network.

At the moment there is not enough time to address that.
